paasta_tools/automatic_rollbacks/signalfx.py

```diff
@@ -0,0 +1,51 @@
+import time
+from typing import Callable
+from typing import Dict
+from typing import List
+
+from signalfx import SignalFx
+from signalfx.signalflow.messages import DataMessage
+from signalfx.signalflow.messages import MetadataMessage
+
+
+def _convert_sfx_timestamp(ts: int) -> float:
+    '''SignalFx uses millisecond int timestamps, we want floating point seconds'''
+    return float(ts) / 1000
+
+
+def tail_signalfx(query: str, lookback_seconds: float, callback: Callable) -> None:
+    if lookback_seconds > 0:
+        start_timestamp_milliseconds = (time.time() - lookback_seconds) * 1000
+    else:
+        start_timestamp_milliseconds = None
+
+    with open('/nail/home/krall/.signalfx_token') as f:
+        sfx_api_token = f.read().strip()
+
+    with SignalFx().signalflow(sfx_api_token) as flow:
+        tsid_metadata_map = {}
+
+        computation = flow.execute(
+            program=query,
+            start=start_timestamp_milliseconds,
+        )
+        for msg in computation.stream():
+            if isinstance(msg, MetadataMessage):
+                # The MetadataMessage occurs before the associated DataMessage.
+                # Extract out the important properties since a lot of it is not important.
+                tsid, props = msg.tsid, msg.properties
+
+                # in SignalFX, query looks like ... .publish('<stream_label>')
+                stream_label: str = props.get('sf_streamLabel', tsid)
+
+                # sf_key seems to have the list of dimensions with a couple extra stuff
+                # The SignalFX query should minimally specify the 'service' dimension.
+                # .sum(by=['dim1','dim2'])
+                sf_key: List[str] = props.get('sf_key', [])
+
+                dimensions: Dict[str, str] = {dim: props[dim] for dim in sf_key}
+                tsid_metadata_map[tsid] = {'stream_label': stream_label, 'dimensions': dimensions}
```

Comment:
Is it guaranteed that we get the Metadata first? Because we use those variables in the elif

Reply:
I'm not sure. Seems like in practice it happens that way. signalfx-python also collects metadata in the same fashion, and doesn't do any sort of error handling for missing metadata: https://github.com/signalfx/signalfx-python/blob/master/signalfx/signalflow/computation.py#L74

If it's a problem, we could queue the data messages until the metadata shows up.